kv: add contending txn_meta for refresh failures

aadityasondhi opened the pull request:
This patch introduces a new field in the error protos for the
transaction retry failures. When a refresh fails due to a write intent
present from another transaction, we now capture its `TxnMeta` and
propagate it up in the `ConflictingTxn` field of the error protos. In
the case of a refresh failure due to a committed value, this field will
be nil, since we clean up write intents after a transaction commits.

This field can be extracted from error messages returned from a
transaction commit operation for observability purposes to identify
the transaction that caused a `RETRY_SERIALIZABLE` error.

Epic: CRDB-26480

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105717)
<!-- Reviewable:end -->

AlexTalks reviewed (changes requested):
I think the comments / changes here are mostly as discussed.  I wanted to avoid leaving a review with the same couple repeated comments throughout, so I just wrote them at the first instances.  Otherwise I think with the minor fixes, completing the remaining TODOs and added test coverage this should be pretty good!

> Reviewed 2 of 26 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbM5Pkg-2kAAOugqTZp:b-x9djii)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @arulajmani)

___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM0rjTA0I5fmOIwpKM):*
nit: `TxnMeta` or transaction identifier
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM11HV3mSeI5iNLgpT):*
nit: "from a transaction commit operation".  `Txn` might be confused for a transaction record, and this is really the error returned by `txn.Commit(..)` or `txn.CommitInBatch(..)`.
___
*[`pkg/kv/mock_transactional_sender.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM1SDU2mjcXQyQJcI7:-NbM1SDU2mjcXQyQJcI8:b-uz54oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/mock_transactional_sender.go#L151)):*
> ```Go
> ) error {
> 	m.txn.Restart(pri, 0 /* upgradePriority */, ts)
> 	return kvpb.NewTransactionRetryWithProtoRefreshError(msg, m.txn.ID, m.txn, nil)
> ```

Replace with helper function as discussed or specify parameter name like `nil /* conflictingTxn */`.  This should be done throughout - I won't repeat this comment everywhere, but it should be done everywhere we end up passing `nil`.
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 808 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM2CNd2pzKMgit7va-:-NbM2CNd2pzKMgit7va0:buh704g) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L808)):*
> ```Go
> 	// different error types are documented above on the metaRestart
> 	// variables.
> 	var contentionTxnMeta *enginepb.TxnMeta
> ```

nit: use `conflictingTxnMeta` to stay consistent
___
*[`pkg/kv/kvpb/api.proto` line 2100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM4xVZ1fXwXPBTijCk:-NbM4xVZ1fXwXPBTijCl:b-worqtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvpb/api.proto#L2100)):*
> ```proto
>   reserved 2;
> 
>   // refresh_from specifies the exclusive lower-bound of the verification. The 
> ```

nit: remove whitespace changes if not necessary
___
*[`pkg/kv/kvpb/errors.go` line 733 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM2n2jEbMRkhhPbmZR:-NbM2n2jEbMRkhhPbmZS:b-og3k1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvpb/errors.go#L733)):*
> ```Go
> 	conflictingTxn *enginepb.TxnMeta,
> ) *TransactionRetryWithProtoRefreshError {
> 	if conflictingTxn == nil {
> ```

As discussed, let's make this field optional rather than storing a zero-value `TxnMeta`. That way this goes away.  Same on the protos, and the other error constructor functions.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
TFTR, please take another look at the changes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbUdlMP-I9hMQ8AmCkw:b-wf6edd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @arulajmani)

___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM0rjTA0I5fmOIwpKM):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: `TxnMeta` or transaction identifier
</blockquote></details>

Done
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM11HV3mSeI5iNLgpT):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: "from a transaction commit operation".  `Txn` might be confused for a transaction record, and this is really the error returned by `txn.Commit(..)` or `txn.CommitInBatch(..)`.
</blockquote></details>

Done
___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 808 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM2CNd2pzKMgit7va-:-NbUcs-i31ifQXREohWm:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L808)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: use `conflictingTxnMeta` to stay consistent
</blockquote></details>

Done
___
*[`pkg/kv/kvpb/api.proto` line 2100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM4xVZ1fXwXPBTijCk:-NbUcvBoDjmgNAcQ2s6m:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvpb/api.proto#L2100)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: remove whitespace changes if not necessary
</blockquote></details>

Done
___
*[`pkg/kv/kvpb/errors.go` line 733 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM2n2jEbMRkhhPbmZR:-NbUd9W6BPBmydKePWu8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/kvpb/errors.go#L733)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

As discussed, let's make this field optional rather than storing a zero-value `TxnMeta`. That way this goes away.  Same on the protos, and the other error constructor functions.
</blockquote></details>

Done.
___
*[`pkg/kv/mock_transactional_sender.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbM1SDU2mjcXQyQJcI7:-NbUdFxr8yvV_qkpyPi9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29eb760eae74d4a36afce1abbbd395ddfce98a90/pkg/kv/mock_transactional_sender.go#L151)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Replace with helper function as discussed or specify parameter name like `nil /* conflictingTxn */`.  This should be done throughout - I won't repeat this comment everywhere, but it should be done everywhere we end up passing `nil`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
Stress tested to make sure it is deterministic:
```
./dev test pkg/kv/kvclient/kvcoord -f TestRefreshFailureIncludesConflictingTxn --stress
...
710 runs so far, 0 failures, over 5m0s
```

aadityasondhi commented:
I noticed I have a bunch of extra files with formatting changes, I will remove those to declutter this PR. It came as a result of using the refactoring command in GoLand.

aadityasondhi committed:
kv: add contending txn_meta for refresh failures

This patch introduces a new field in the error protos for the
transaction retry failures. When a refresh fails due to a write intent
present from another transaction, we now capture its `TxnMeta` and
propagate it up in the `ConflictingTxn` field of the error protos. In
the case of a refresh failure due to a committed value, this field will
be nil, since we clean up write intents after a transaction commits.

This field can be extracted from error messages returned from a
transaction commit operation for observability purposes to identify
the transaction that caused a `RETRY_SERIALIZABLE` error.

Epic: CRDB-26480

Release note: None

arulajmani reviewed (commented):
Exciting to see this come together! I mostly have minor comments, but this should be good to go once they're addressed. 

> Reviewed 5 of 25 files at r3, 14 of 14 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-Nb_5bzgB17FooyZCus5:b-5gc7cy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @AlexTalks)

___
*[`-- commits` line 5 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbZzpfxD4ovnQRfb7Qq):*
Consider adding some words here about when this field is not populated. In particular, when the refresh fails due to a committed value (and not a conflicting write intent). 
___
*[`pkg/kv/kvpb/errors.go` line 736 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_08SP4snGfZ6qxiG5:-Nb_08SP4snGfZ6qxiG6:b1zriqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L736)):*
> ```Go
> // NewTransactionRetryWithProtoRefreshError but with a conflicting txn if
> // available.
> func NewTransactionRetryWithProtoRefreshErrorWithConflictingTxn(
> ```

Do we need this new constructor? Could we instead expand `NewTransactionRetryErrorWithProtoRefreshError` to include an optional `conflictingTxn` parameter instead?
___
*[`pkg/kv/kvpb/errors.go` line 805 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_0Vm21Gyfe67zPobu:-Nb_0Vm21Gyfe67zPobv:b-auib5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L805)):*
> ```Go
> // NewTransactionRetryErrorWithConflictingTxn initializes a new
> // TransactionRetryError with a conflicting txn included.
> func NewTransactionRetryErrorWithConflictingTxn(
> ```

Same comment as above, could we get rid of this function by passing in an optional `conflictingTxn` to `NewTransactionRetryError` instead?
___
*[`pkg/kv/kvpb/errors.go` line 1474 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1A6G113khAGhFDgN:-Nb_1A6G113khAGhFDgO:benpsw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L1474)):*
> ```Go
> // includes a *enginepb.TxnMeta for a conflicting transaction if its write
> // intent is found.
> func NewRefreshFailedErrorWithConflictingTxn(
> ```

Same comment as above.
___
*[`pkg/kv/kvpb/errors.go` line 1480 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1lmM2i2LhdWTq1SC:-Nb_1lmM2i2LhdWTq1SD:bwtwxck) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L1480)):*
> ```Go
> 	conflictingTxn *enginepb.TxnMeta,
> ) *RefreshFailedError {
> 	return &RefreshFailedError{
> ```

Should we assert `conflictingTxn` is not nil if the refresh failed reason is `REASON_INTENT`? Maybe even the other way around as well?
___
*[`pkg/kv/kvpb/errors.proto` line 496 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1Fc02owwC-r-uUnV:-Nb_1Fc02owwC-r-uUnW:b-9ks4sv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.proto#L496)):*
> ```proto
>   optional string msg_redactable = 4 [(gogoproto.nullable) = false, (gogoproto.customtype) = "github.com/cockroachdb/redact.RedactableString"];
> 
>   // The conflicting txn identifier (if available).
> ```

Should we expand this comment a bit more? It might be worth adding more words about when this field is set and when it isn't.
___
*[`pkg/kv/kvpb/errors.proto` line 629 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1UVb7jTYBxEwInCY:-Nb_1UVb7jTYBxEwInCZ:b-ooszz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.proto#L629)):*
> ```proto
>   optional util.hlc.Timestamp timestamp = 3 [(gogoproto.nullable) = false];
>   
>   // The conflicting txn identifier (if available).
> ```

This could use more words as well.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4299 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2SMjELSk5g17HpCG:-Nb_2SMjELSk5g17HpCH:bdkaztq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4299)):*
> ```Go
> 		committed    bool
> 		rangeRefresh bool
> 		key1         string
> ```

nit: consider pulling key1 and key2 into the test itself, as these aren't changing subtest to subtest.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4304 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_4X0o26LlqFIvdclp:-Nb_4X0o26LlqFIvdclq:bu7gnwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4304)):*
> ```Go
> 		{
> 			name:         "write intent found in single key refresh",
> 			committed:    false,
> ```

If you take my suggestion about pulling key1 and key2 into the subtest, you could write this test entirely by using `testutils.RunTrueAndFalse` -- that way, you don't need this tc struct. Something like:

```go
testutils.RunTrueAndFalse(t, "committed", func (t * testing.T, committed bool) {  
 testutils.RunTrueAndFalse(t, "range-refresh", func (t *testing.T, rangeRefresh bool) {  
  ...    
 })  
})
```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2zN9APYOFgdbW5Mb:-Nb_2zN9APYOFgdbW5Mc:beg8m9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4350)):*
> ```Go
> 			}
> 
> 			_, err = txn2.Get(ctx, tc.key2)
> ```

Consider adding a comment explaining this step. Something about bumping the timestamp cache on key2 so that txn1's put causes its write timestamp to skew, thereby necessitating a refresh.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2n-G5cSyRa-nhnFT:-Nb_2n-G5cSyRa-nhnFU:bi35zxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4358)):*
> ```Go
> 			if tc.committed {
> 				require.NoError(t, txn2.Commit(ctx))
> 				// force intent clean up
> ```

nit: s/force intent clean up/Force intent cleanup./g


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
TFTR, will do the suggested refactoring and post an update here once this is ready for another look.

aadityasondhi reviewed (commented):
I ended up taking all suggestions, they helped clean up the code quite a bit!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbuEq0qBDqnZLour0nE:b-tkflei)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @arulajmani)

___
*[`-- commits` line 5 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbZzpfxD4ovnQRfb7Qq):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding some words here about when this field is not populated. In particular, when the refresh fails due to a committed value (and not a conflicting write intent). 
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 736 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_08SP4snGfZ6qxiG5:-Nbu9MLc-hZtOE5QySSM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L736)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we need this new constructor? Could we instead expand `NewTransactionRetryErrorWithProtoRefreshError` to include an optional `conflictingTxn` parameter instead?
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 805 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_0Vm21Gyfe67zPobu:-Nbu9PTj84tEmluHVL7p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L805)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Same comment as above, could we get rid of this function by passing in an optional `conflictingTxn` to `NewTransactionRetryError` instead?
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 1474 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1A6G113khAGhFDgN:-Nbu9QBh69DcHcQxYEnx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L1474)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 1480 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1lmM2i2LhdWTq1SC:-Nbu9S206Vy_fEXVQ28N:b-ljs39v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L1480)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we assert `conflictingTxn` is not nil if the refresh failed reason is `REASON_INTENT`? Maybe even the other way around as well?
</blockquote></details>

Done. I ended up using `log.Error()` for this assertion.
___
*[`pkg/kv/kvpb/errors.proto` line 496 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1Fc02owwC-r-uUnV:-NbuBJMZ5FacdJ6GECEv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.proto#L496)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we expand this comment a bit more? It might be worth adding more words about when this field is set and when it isn't.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.proto` line 629 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_1UVb7jTYBxEwInCY:-NbuBKER8Zc6-G3pb4G3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.proto#L629)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This could use more words as well.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4299 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2SMjELSk5g17HpCG:-Nbu9dEu7JEx_tUJkc9C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4299)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider pulling key1 and key2 into the test itself, as these aren't changing subtest to subtest.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4304 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_4X0o26LlqFIvdclp:-Nbu9fFS2h53suznZ9Al:b-gvy3a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4304)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

If you take my suggestion about pulling key1 and key2 into the subtest, you could write this test entirely by using `testutils.RunTrueAndFalse` -- that way, you don't need this tc struct. Something like:

```go
testutils.RunTrueAndFalse(t, "committed", func (t * testing.T, committed bool) {  
 testutils.RunTrueAndFalse(t, "range-refresh", func (t *testing.T, rangeRefresh bool) {  
  ...    
 })  
})
```
</blockquote></details>

Thanks, this is pretty neat!
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4350 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2zN9APYOFgdbW5Mb:-Nbu9hax2xaSDRHNlFoO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4350)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding a comment explaining this step. Something about bumping the timestamp cache on key2 so that txn1's put causes its write timestamp to skew, thereby necessitating a refresh.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_2n-G5cSyRa-nhnFT:-Nbu9iPYCX5dldYFwtlD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4358)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/force intent clean up/Force intent cleanup./g
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (changes requested):
Looks great!  I have 2 very minor comments; in particular once we address the assertion I am ready to approve!

> Reviewed 7 of 25 files at r3, 7 of 7 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbwL9Am1Uis1zuWD5H0:bwmr8n3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 852 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwFdDcBdtK2jthi98L:-NbwFdDcBdtK2jthi98M:b633gkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L852)):*
> ```Go
> 	retErr := kvpb.NewTransactionRetryWithProtoRefreshError(
> 		redact.Sprint(pErr),
> 		errTxnID,
> ```

One downside of using GoLand's refactoring tool is that it gets rid of comments unfortunately.  Can we re-add the comment?
___
*[`pkg/kv/kvpb/errors.go` line 1483 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwIiFkAl7f-ObAdAoW:-NbwIiFkAl7f-ObAdAoX:bggb9bw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L1483)):*
> ```Go
> 	if (reason == RefreshFailedError_REASON_INTENT && options.conflictingTxn == nil) ||
> 		(reason != RefreshFailedError_REASON_INTENT && options.conflictingTxn != nil) {
> 		log.Error(context.TODO(), "conflictingTxn should be non-nil if and only if refresh failed with REASON_INTENT")
> ```

Shouldn't this actually be a `panic` or other form of assertion? I feel like if there's some way of failing a refresh on a found intent but not actually capturing the conflicting `TxnMeta`, we'd like to catch that in a failed test, not just log it.  Additionally I'm not sure we'd want to use `contxt.TODO()` a place like this.

Also, a very small nit: this may be more readable if you separated out the two invalid cases (intent + missing txnmeta, non-intent + valid txnmeta); I'll leave this up to you however.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4304 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_4X0o26LlqFIvdclp:-NbwK2pRCWuW4V1f-w69:bm5dbxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4304)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Thanks, this is pretty neat!
</blockquote></details>

Agreed, this looks great!
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4344 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwKFr7CY0ORCiDhESR:-NbwKFr7CY0ORCiDhESS:bj84xhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4344)):*
> ```Go
> 				require.NotNil(t, tErr.ConflictingTxn)
> 				require.Equal(t, txn2.ID(), tErr.ConflictingTxn.ID)
> 				require.Equal(t, int32(1), tErr.ConflictingTxn.CoordinatorNodeID)
> ```

Thanks! :smile: 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
This is looking quite nice. I have a few nits that we should address before merging, but other than that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbylbhyDE4G57GwHemv:bljfrkg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi and @AlexTalks)

___
*[`pkg/kv/kvpb/errors.go` line 736 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nb_08SP4snGfZ6qxiG5:-NbydJ402x-2cSY73aFy:b-kywhyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afda0a60ce7b59a7da04f7e8bb8248ec0706f11/pkg/kv/kvpb/errors.go#L736)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This turned out quite nice! Thanks for refactoring.
___
*[`pkg/kv/kvpb/errors.go` line 722 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbymFXaB3wU0LU8s8K4:-NbymFXbChL4xEc69d8X:b3nx2wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L722)):*
> ```Go
> }
> 
> // RetryErrOption is used to annotate optional fields in retry related errors.
> ```

Slightly curious, are we planning on using this for errors other than `RefreshFailedErrors` in the near term?
___
*[`pkg/kv/kvpb/errors.go` line 733 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbycnNV8EG2cyTRdNXA:-NbycnNV8EG2cyTRdNXB:b7q63gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L733)):*
> ```Go
> }
> 
> func ErrorWithConflictingTxn(txn *enginepb.TxnMeta) RetryErrOption {
> ```

nit: s/ErrorWithConflictingTxn/WithConflictingTxn/g


Let's add a comment here explaining what this option is all about.

----
side note: I'm surprised the linter didn't complain here, given this is a public function.
___
*[`pkg/kv/kvpb/errors.go` line 739 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbym6fZD8O7HeLOfECu:-Nbym6fZD8O7HeLOfECv:b5j46xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L739)):*
> ```Go
> }
> 
> // NewTransactionRetryWithProtoRefreshError initializes a new TransactionRetryWithProtoRefreshError.
> ```

While you're here, would you mind wrapping this comment to 80 chars?
___
*[`pkg/kv/kvpb/errors.go` line 1483 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwIiFkAl7f-ObAdAoW:-NbygRjC9KE5yams8fiJ:b-605gng) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L1483)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Shouldn't this actually be a `panic` or other form of assertion? I feel like if there's some way of failing a refresh on a found intent but not actually capturing the conflicting `TxnMeta`, we'd like to catch that in a failed test, not just log it.  Additionally I'm not sure we'd want to use `contxt.TODO()` a place like this.

Also, a very small nit: this may be more readable if you separated out the two invalid cases (intent + missing txnmeta, non-intent + valid txnmeta); I'll leave this up to you however.
</blockquote></details>

+1 on the `context.TODO()` usage here. We should very rarely, if ever, be using that in non-test code. Let's thread in the real thing.
___
*[`pkg/kv/kvpb/errors.proto` line 267 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbyi41BATbO6deKn56w:-Nbyi41BATbO6deKn56x:bstj4bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L267)):*
> ```proto
>   optional string extra_msg = 2 [(gogoproto.nullable) = false];
>   optional string extra_msg_redactable = 3 [(gogoproto.nullable) = false, (gogoproto.customtype) = "github.com/cockroachdb/redact.RedactableString"];
>   // The identifier for a conflicting transaction. This is bubbled up from a
> ```

nit: Should this say "conflicting transaction's meta"? identifier generally refers to the ID of the transaction, which is a field on the `TxnMeta`.
___
*[`pkg/kv/kvpb/errors.proto` line 268 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbyjfeh7j5SBzDnWdRL:-Nbyjfeh7j5SBzDnWdRM:bz8yqj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L268)):*
> ```proto
>   optional string extra_msg_redactable = 3 [(gogoproto.nullable) = false, (gogoproto.customtype) = "github.com/cockroachdb/redact.RedactableString"];
>   // The identifier for a conflicting transaction. This is bubbled up from a
>   // RefreshFailedError.
> ```

nit: Should we add a bit more words about when this field isn't set? Maybe something like "This field is only set if the retry error is constructed in response to a RefreshFailedError. In all other cases, or if the RefreshFailedError does not contain conflicting txn information, this field is unset." 
___
*[`pkg/kv/kvpb/errors.proto` line 498 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbykkql6Ohqfc4hTFAo:-Nbykkql6Ohqfc4hTFAp:bqefw2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L498)):*
> ```proto
>   optional string msg_redactable = 4 [(gogoproto.nullable) = false, (gogoproto.customtype) = "github.com/cockroachdb/redact.RedactableString"];
> 
>   // The conflicting transaction identifier. This field is bubbled up from a
> ```

s/identifier/meta/g

nit: "This field is only set when a TransactionRetryWithProtoRefreshError is constructed in response to a RefreshFailedError which includes information about the conflicting transaction that caused the refresh to fail. In all other cases this field is unset".
___
*[`pkg/kv/kvpb/errors.proto` line 634 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbykKdxCZTRAwJp5-bM:-NbykKdxCZTRAwJp5-bN:b553b95) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L634)):*
> ```proto
>   // The conflicting txn identifier, if available. This field is used when the
>   // refresh fails with `REASON_INTENT` and is bubbled up for observability
>   // purposes to track the conflicting transaction.
> ```

nit: Let's add "This field is not set if the refresh fails because of a committed value" at the end.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4311 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbyhN4I3xDsBRTpjDZr:-NbyhN4JAvZnDClDa-_I:b-b8l49) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4311)):*
> ```Go
> 				require.NoError(t, err)
> 			} else {
> 				_, err = txn1.Get(ctx, "a")
> ```

nit: you could define `keyA := "a"` and `keyB := "b"` so that we're not re-using raw strings in a bunch of places.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Thanks for the reviews, resolved all comments/nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-Nbz8u4X3sjJ8TZz2B0-:brmw6ax)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/txn_coord_sender.go` line 852 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwFdDcBdtK2jthi98L:-Nbz1MUC1TellmvwL3k9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L852)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

One downside of using GoLand's refactoring tool is that it gets rid of comments unfortunately.  Can we re-add the comment?
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 722 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbymFXaB3wU0LU8s8K4:-Nbz1N_y6VgIcgVIAkXg:b-n5fae0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L722)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Slightly curious, are we planning on using this for errors other than `RefreshFailedErrors` in the near term?
</blockquote></details>

I don't think so. 
___
*[`pkg/kv/kvpb/errors.go` line 733 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbycnNV8EG2cyTRdNXA:-Nbz1fN22omQOHyAeyO0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L733)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/ErrorWithConflictingTxn/WithConflictingTxn/g


Let's add a comment here explaining what this option is all about.

----
side note: I'm surprised the linter didn't complain here, given this is a public function.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 739 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbym6fZD8O7HeLOfECu:-Nbz2Vf-2vLBwTik5b18:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L739)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

While you're here, would you mind wrapping this comment to 80 chars?
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.go` line 1483 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbwIiFkAl7f-ObAdAoW:-Nbz2ZkUCkvHrXpR75eT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.go#L1483)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

+1 on the `context.TODO()` usage here. We should very rarely, if ever, be using that in non-test code. Let's thread in the real thing.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.proto` line 267 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbyi41BATbO6deKn56w:-Nbz83crCUItGA1njlzW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L267)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Should this say "conflicting transaction's meta"? identifier generally refers to the ID of the transaction, which is a field on the `TxnMeta`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.proto` line 268 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbyjfeh7j5SBzDnWdRL:-Nbz84SNBg1Qw2aE2I7s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L268)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Should we add a bit more words about when this field isn't set? Maybe something like "This field is only set if the retry error is constructed in response to a RefreshFailedError. In all other cases, or if the RefreshFailedError does not contain conflicting txn information, this field is unset."
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.proto` line 498 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-Nbykkql6Ohqfc4hTFAo:-Nbz857RAgNluJWMfLfZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L498)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/identifier/meta/g

nit: "This field is only set when a TransactionRetryWithProtoRefreshError is constructed in response to a RefreshFailedError which includes information about the conflicting transaction that caused the refresh to fail. In all other cases this field is unset".
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/errors.proto` line 634 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbykKdxCZTRAwJp5-bM:-Nbz8N5bE12D7p6-niYy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvpb/errors.proto#L634)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Let's add "This field is not set if the refresh fails because of a committed value" at the end.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_server_test.go` line 4311 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbyhN4I3xDsBRTpjDZr:-Nbz4f8S5uBGne-JaGYB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2dd6b1b926f94e1c8af269ba21e2b54c78887ca/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L4311)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you could define `keyA := "a"` and `keyB := "b"` so that we're not re-using raw strings in a bunch of places.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Some tiny nits, but I am ready to approve this as it's looking good, nice work!

> Reviewed 9 of 9 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-NbzKevf4wek6mHB80_b:b-xk1594)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi and @arulajmani)

___
*[`pkg/kv/kvpb/errors.go` line 1490 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbzJyGY0uugJjeN4fyq:-NbzJyGY0uugJjeN4fyr:bowcpbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233614cd0cd5a1a5c29d3616ce117124bb2e748/pkg/kv/kvpb/errors.go#L1490)):*
> ```Go
> 	}
> 	if reason == RefreshFailedError_REASON_INTENT && options.conflictingTxn == nil {
> 		log.Fatal(ctx, "conflictingTxn should be set if refresh failed with REASON_INTENT")
> ```

nit: I think either `log.Fatal` or even `panic` (which doesn't need a `ctx`) can work here.  We do use `panic` in various places when things should never happen, like an unhandled enum value in a switch.  Leaving which one you use here up to you.
___
*[`pkg/kv/kvpb/errors.proto` line 639 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbzKVnr7-y7ppihiGA5:-NbzKVnr7-y7ppihiGA6:b-u3m95h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233614cd0cd5a1a5c29d3616ce117124bb2e748/pkg/kv/kvpb/errors.proto#L639)):*
> ```proto
>   optional util.hlc.Timestamp timestamp = 3 [(gogoproto.nullable) = false];
>   
>   // The conflicting txn identifier, if available. This field is used when the
> ```

nit: You should probably keep this in sync with what you wrote above.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
Thanks for the final review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-:-Nbzt16pAKwvJGdnXcl2:b-picocs)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @AlexTalks and @arulajmani)

___
*[`pkg/kv/kvpb/errors.go` line 1490 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105717#-NbzJyGY0uugJjeN4fyq:-NbzsM62ECq9Epf14Ifb:bdb7pa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e233614cd0cd5a1a5c29d3616ce117124bb2e748/pkg/kv/kvpb/errors.go#L1490)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: I think either `log.Fatal` or even `panic` (which doesn't need a `ctx`) can work here.  We do use `panic` in various places when things should never happen, like an unhandled enum value in a switch.  Leaving which one you use here up to you.
</blockquote></details>

I'll just keep this. But thanks for the context!


<!-- Sent from Reviewable.io -->

aadityasondhi commented:
bors r=AlexTalks,arulajmani
